<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Genotype-Phenotype Integration for Antimicrobial Resistance Research • AMRgen</title>
<script src="lightswitch.js"></script><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Genotype-Phenotype Integration for Antimicrobial Resistance Research">
<meta name="description" content="The AMRgen package provides tools for interpreting antimicrobial resistance (AMR) genes, integrating genomic data with phenotypic antimicrobial susceptibility testing (AST) data, and calculating genotype-phenotype associations. This package extends the functionality of the AMR package, enabling researchers to fetch MIC distribution data from EUCAST, visualise multidrug resistance patterns using UpSet plots, and generate NCBI-compliant antibiograms. Designed for AMR research and surveillance, AMRgen aims to support data-driven decision-making in combating AMR on a global scale.">
<meta property="og:description" content="The AMRgen package provides tools for interpreting antimicrobial resistance (AMR) genes, integrating genomic data with phenotypic antimicrobial susceptibility testing (AST) data, and calculating genotype-phenotype associations. This package extends the functionality of the AMR package, enabling researchers to fetch MIC distribution data from EUCAST, visualise multidrug resistance patterns using UpSet plots, and generate NCBI-compliant antibiograms. Designed for AMR research and surveillance, AMRgen aims to support data-driven decision-making in combating AMR on a global scale.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">AMRgen</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/AMRgen.html">Get Started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/AMRgen.html">Introduction</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/interpretAMR/AMRgen"><span class="fa fab fa-github"></span> Source Code</a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="amrgen">AMRgen<a class="anchor" aria-label="anchor" href="#amrgen"></a>
</h1></div>
<p>R Package for Genetic and Phenotypic Resistance Interpretation</p>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p><strong>AMRgen</strong> is an open-source R package designed to bridge the gap between genotypic and phenotypic antimicrobial resistance (AMR) data. Developed as an extension to the <a href="https://github.com/msberends/AMR" class="external-link">AMR R package</a>, it provides tools to interpret AMR genes, integrate these findings with antimicrobial susceptibility test (AST) data, and calculate genotype-phenotype associations.</p>
<p>This package is developed in collaboration with the ESGEM-AMR Working Group and is tailored for researchers and healthcare professionals tackling AMR globally.</p>
<hr>
</div>
<div class="section level2">
<h2 id="key-features">Key Features<a class="anchor" aria-label="anchor" href="#key-features"></a>
</h2>
<ul>
<li>
<strong>Import Genotype and Phenotype Data</strong>: Import from common formats (NCBI AST for phenotypes; AMRfinderplus and hAMRonization for genotypes</li>
<li>
<strong>Genotype-Phenotype Integration</strong>: Links AMR gene presence with phenotypic resistance profiles, enabling deeper insights into resistance mechanisms.</li>
<li>
<strong>Automated EUCAST MIC Distribution Integration</strong>: Fetch MIC distribution data directly from <a href="https://mic.eucast.org" class="external-link">EUCAST</a> for seamless comparison with local susceptibility data.</li>
<li>
<strong>Visualisation</strong>: Generate powerful UpSet plots to identify intersections of AMR gene presence and phenotypic resistance, highlighting multidrug resistance patterns.</li>
<li>
<strong>Modular and Extensible</strong>: Leverages the robust foundation of the AMR package, including antibiotic selectors and clinical breakpoint interpretations.</li>
</ul>
<blockquote>
<p>Planned for development - <strong>NCBI-Compliant Export</strong>: Export phenotype data to NCBI-compliant antibiogram format. - <strong>Expanded Data Import</strong>: Import and parse phenotype data from other tools (e.g. CARD, ResFinder).</p>
</blockquote>
<hr>
</div>
<div class="section level2">
<h2 id="getting-started">Getting Started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>To install and explore the package, follow the instructions below:</p>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>Note that this package requires the latest version of the <code>AMR</code> package (still in beta).</p>
<p>Install the latest version of the <code>AMR</code> package with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span> <span class="co"># if you haven't already</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"msberends/AMR"</span><span class="op">)</span></span></code></pre></div>
<p>Then install this package</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install from GitHub</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"interpretAMR/AMRgen"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="usage-examples">Usage Examples<a class="anchor" aria-label="anchor" href="#usage-examples"></a>
</h2>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://interpretAMR.github.io/AMRgen/" class="external-link">AMRgen</a></span><span class="op">)</span></span></code></pre>
<div class="section level3">
<h3 id="import-pheno-data-from-ncbi-ast-and-geno-data-amrfinderplus-output-and-compare-genopheno-for-drugs-of-interest">Import pheno data (from NCBI AST) and geno data (AMRfinderplus output), and compare geno/pheno for drugs of interest<a class="anchor" aria-label="anchor" href="#import-pheno-data-from-ncbi-ast-and-geno-data-amrfinderplus-output-and-compare-genopheno-for-drugs-of-interest"></a>
</h3>
<pre><code><span><span class="co"># import small example E. coli AST data from NCBI (without re-interpreting resistance)</span></span>
<span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/import_ncbi_ast.html">import_ncbi_ast</a></span><span class="op">(</span><span class="st">"testdata/Ecoli_AST_NCBI_n50.tsv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># import small example E. coli AMRfinderplus data</span></span>
<span><span class="va">geno</span> <span class="op">&lt;-</span> <span class="fu">import_amrfp</span><span class="op">(</span><span class="st">"testdata/Ecoli_AMRfinderplus_n50.tsv"</span>, <span class="st">"Name"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get matrix combining data on phenotype (binary R and NWT for a drug) and genotype (binary presence/absence for markers for relevant drug class/es)</span></span>
<span></span>
<span><span class="co"># 1: meropenem phenotype vs carbapenem/cephalosporin genetic markers</span></span>
<span><span class="va">mero_vs_blaGenes</span> <span class="op">&lt;-</span> <span class="fu">get_binary_matrix</span><span class="op">(</span><span class="va">geno</span>, <span class="va">pheno</span>, antibiotic<span class="op">=</span><span class="st">"Meropenem"</span>, drug_class_list<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Carbapenems"</span>, <span class="st">"Cephalosporins"</span><span class="op">)</span>, sir_col<span class="op">=</span><span class="st">"Resistance phenotype"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 2: ciprofloxacin phenotype vs quinolone genetic markers</span></span>
<span><span class="va">cipro_vs_quinoloneMarkers</span> <span class="op">&lt;-</span> <span class="fu">get_binary_matrix</span><span class="op">(</span><span class="va">geno</span>, <span class="va">pheno</span>, <span class="st">"Ciprofloxacin"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Quinolones"</span><span class="op">)</span>, sir_col<span class="op">=</span><span class="st">"Resistance phenotype"</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="import-small-example-e-coli-ast-data-from-ncbi-and-re-interpret-resistance-and-ecoff-using-amr-package">Import small example E. coli AST data from NCBI, and re-interpret resistance and ECOFF using AMR package<a class="anchor" aria-label="anchor" href="#import-small-example-e-coli-ast-data-from-ncbi-and-re-interpret-resistance-and-ecoff-using-amr-package"></a>
</h3>
<p><em><strong>(WARNING: phenotype interpretation can take a few minutes)</strong></em></p>
<pre><code><span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/import_ncbi_ast.html">import_ncbi_ast</a></span><span class="op">(</span><span class="st">"testdata/Ecoli_AST_NCBI_n50.tsv"</span>, interpret <span class="op">=</span> <span class="cn">T</span>, ecoff<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="investigate-ciprofloxacin-resistance-vs-quinolone-genotype-markers-via-solo-ppv-and-upset-plots">Investigate ciprofloxacin resistance vs quinolone genotype markers, via solo PPV and upset plots<a class="anchor" aria-label="anchor" href="#investigate-ciprofloxacin-resistance-vs-quinolone-genotype-markers-via-solo-ppv-and-upset-plots"></a>
</h3>
<p><em>(Using the resistance interpretations as imported from NCBI AST file, not reinterpreted from assay values)</em></p>
<pre><code><span><span class="co"># import larger example E. coli AST data from NCBI (without re-interpreting resistance)</span></span>
<span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/import_ncbi_ast.html">import_ncbi_ast</a></span><span class="op">(</span><span class="st">"testdata/Ecoli_AST_NCBI.tsv.gz"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># import larger example E. coli AMRfinderplus data</span></span>
<span><span class="va">geno</span> <span class="op">&lt;-</span> <span class="fu">import_amrfp</span><span class="op">(</span><span class="st">"testdata/Ecoli_AMRfinderplus.tsv.gz"</span>, <span class="st">"Name"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># find genomes with just one quinolone resistance marker, then estimate and plot positive predictive value (PPV) for ciprofloxacin resistance/NWT</span></span>
<span><span class="va">soloPPV_cipro</span> <span class="op">&lt;-</span> <span class="fu">solo_ppv_analysis</span><span class="op">(</span><span class="va">geno</span>, <span class="va">pheno</span>, antibiotic<span class="op">=</span><span class="st">"Ciprofloxacin"</span>, drug_class_list<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Quinolones"</span><span class="op">)</span>, sir_col<span class="op">=</span><span class="st">"Resistance phenotype"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># view PPV summary statistics:</span></span>
<span><span class="va">soloPPV_cipro</span><span class="op">$</span><span class="va">solo_stats</span></span>
<span></span>
<span><span class="co"># plot PPV summary:</span></span>
<span><span class="va">soloPPV_cipro</span><span class="op">$</span><span class="va">combined_plot</span></span>
<span></span>
<span><span class="co"># get matrix combining data on ciprofloxacin phenotype (MIC, plus binary R and NWT) and genotype (binary presence/absence for quinolone resistance markers)</span></span>
<span><span class="va">cip_bin</span><span class="op">&lt;-</span> <span class="fu">get_binary_matrix</span><span class="op">(</span><span class="va">geno</span>, <span class="va">pheno</span>, antibiotic<span class="op">=</span><span class="st">"Ciprofloxacin"</span>, drug_class_list<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Quinolones"</span><span class="op">)</span>, sir_col<span class="op">=</span><span class="st">"Resistance phenotype"</span>, keep_assay_values<span class="op">=</span><span class="cn">T</span>, keep_assay_values_from <span class="op">=</span> <span class="st">"mic"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># do upset plot of MIC vs genotype marker combinations (using AMRgen function, not requiring complexUpset)</span></span>
<span><span class="fu">amr_upset</span><span class="op">(</span><span class="va">cip_bin</span>, min_set_size<span class="op">=</span><span class="fl">2</span>, order<span class="op">=</span><span class="st">"mic"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># do upset plot of MIC vs genotype marker combinations (using complexUpset)</span></span>
<span><span class="fu">amr_complexUpset</span><span class="op">(</span><span class="va">cip_bin</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="investigate-ciprofloxacin-resistance-vs-quinolone-genotype-markers-via-solo-ppv-and-upset-plots-1">Investigate ciprofloxacin resistance vs quinolone genotype markers, via solo PPV and upset plots<a class="anchor" aria-label="anchor" href="#investigate-ciprofloxacin-resistance-vs-quinolone-genotype-markers-via-solo-ppv-and-upset-plots-1"></a>
</h3>
<p><em>(Using phenotypes reinterpreted from assay values in the NCBI AST file; note these have been pre-computed to save time for this example)</em></p>
<pre><code><span><span class="co"># import E. coli AST data from NCBI (R and NWT variables have been pre-computed from the raw NCBI AST file using: import_ncbi_ast("testdata/Ecoli_AST_NCBI.tsv", interpret=T, ecoff=T)</span></span>
<span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu">read_tsv</span><span class="op">(</span><span class="st">"testdata/Ecoli_AST_NCBI_reinterpreted.tsv.gz"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>drug_agent<span class="op">=</span><span class="fu">as.ab</span><span class="op">(</span><span class="va">drug_agent</span><span class="op">)</span>, spp_pheno<span class="op">=</span><span class="fu">as.mo</span><span class="op">(</span><span class="va">spp_pheno</span><span class="op">)</span>, mic<span class="op">=</span><span class="fu">as.mic</span><span class="op">(</span><span class="va">mic</span><span class="op">)</span>, disk<span class="op">=</span><span class="fu">as.disk</span><span class="op">(</span><span class="va">disk</span><span class="op">)</span>, pheno<span class="op">=</span><span class="fu">as.sir</span><span class="op">(</span><span class="va">pheno</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># import AMRfinderplus data</span></span>
<span><span class="va">geno</span> <span class="op">&lt;-</span> <span class="fu">import_amrfp</span><span class="op">(</span><span class="st">"testdata/Ecoli_AMRfinderplus.tsv.gz"</span>, <span class="st">"Name"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># do positive predictive value (PPV) analysis for quinolone resistance markers vs ciprofloxacin resistance/NWT</span></span>
<span><span class="va">soloPPV_cipro</span> <span class="op">&lt;-</span> <span class="fu">solo_ppv_analysis</span><span class="op">(</span><span class="va">geno</span>, <span class="va">pheno</span>, antibiotic<span class="op">=</span><span class="st">"Ciprofloxacin"</span>, drug_class_list<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Quinolones"</span><span class="op">)</span>, sir_col<span class="op">=</span><span class="st">"Resistance phenotype"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">soloPPV_cipro</span><span class="op">$</span><span class="va">solo_stats</span></span>
<span></span>
<span><span class="va">soloPPV_cipro</span><span class="op">$</span><span class="va">combined_plot</span></span>
<span></span>
<span><span class="co"># get matrix combining data on ciprofloxacin phenotype (MIC, plus binary R and NWT) and genotype (binary presence/absence for quinolone resistance markers)</span></span>
<span><span class="va">cip_bin</span><span class="op">&lt;-</span> <span class="fu">get_binary_matrix</span><span class="op">(</span><span class="va">geno</span>, <span class="va">pheno</span>, antibiotic<span class="op">=</span><span class="st">"Ciprofloxacin"</span>, drug_class_list<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Quinolones"</span><span class="op">)</span>, sir_col<span class="op">=</span><span class="st">"Resistance phenotype"</span>, keep_assay_values<span class="op">=</span><span class="cn">T</span>, keep_assay_values_from <span class="op">=</span> <span class="st">"mic"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># do upset plot of MIC vs genotype marker combinations (using AMRgen function, not requiring complexUpset)</span></span>
<span><span class="fu">amr_upset</span><span class="op">(</span><span class="va">cip_bin</span>, min_set_size<span class="op">=</span><span class="fl">2</span>, order<span class="op">=</span><span class="st">"mic"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># do upset plot of MIC vs genotype marker combinations (using complexUpset)</span></span>
<span><span class="fu">amr_complexUpset</span><span class="op">(</span><span class="va">cip_bin</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="contributions">Contributions<a class="anchor" aria-label="anchor" href="#contributions"></a>
</h2>
<p>Contributions are welcome! If you encounter issues or wish to suggest new features, please open an issue or submit a pull request.</p>
</div>
<div class="section level2">
<h2 id="licence">Licence<a class="anchor" aria-label="anchor" href="#licence"></a>
</h2>
<p>This package is distributed under the GNU GPL-3.0 Licence. See <code>LICENSE</code> for details.</p>
</div>
<div class="section level2">
<h2 id="dev-notes">Dev notes:<a class="anchor" aria-label="anchor" href="#dev-notes"></a>
</h2>
<p>Comnon formats generated by import functions, and expected by analysis/plotting functions</p>
<div class="section level3">
<h3 id="definition-genotype-dataframe">Definition: ‘genotype’ dataframe<a class="anchor" aria-label="anchor" href="#definition-genotype-dataframe"></a>
</h3>
<p>required fields: - a column indicating the sample ID (any string; to be matched to phenotype data) - a column named ‘marker’ indicating the label for a specific gene or mutation (S3 class ‘gene’)</p>
<p>at least one of: - a column named ‘drug_agent’ (S3 class ‘ab’; this will usually be generated by a genotype parser function applying as.ab) - a column named ‘drug_class’ indicating a drug class associated with this marker (controlled vocab string; allowed values are those in antbiotics$group, this will usually be generated by a genotype parser function, if not provided it will be generated from ‘drug_agent’) [might develop S3 class? e.g. needs to include ‘efflux’]</p>
<p>optionally: - a column indicating the species (S3 class mo; to facilitate interpretation)</p>
<p>genotype parsers should be generating all 4 fields</p>
</div>
<div class="section level3">
<h3 id="definition-phenotype-dataframe">Definition: ‘phenotype’ dataframe<a class="anchor" aria-label="anchor" href="#definition-phenotype-dataframe"></a>
</h3>
<p>required fields: - a column indicating the sample ID (any string; to be matched to genotype data) - a column named ‘drug_agent’ (S3 class ‘ab’; this will usually be generated by a phenotype parser function applying as.ab)</p>
<p>at least one of: - a column of class ‘disk’ or ‘mic’ - a column of class ‘sir’</p>
<p>optionally: - a column indicating the species (S3 class mo; to facilitate interpretation)</p>
</div>
<div class="section level3">
<h3 id="expected-workflow-target-for-dev">Expected workflow (target for dev)<a class="anchor" aria-label="anchor" href="#expected-workflow-target-for-dev"></a>
</h3>
<ul>
<li>import genotype data -&gt; genotype dataframe (e.g. <code>import_amrfp</code>)</li>
<li>import phenotype data -&gt; phenotype dataframe (e.g. <code>import_ncbi_ast</code>)
<ul>
<li>interpret SIR if required (as.sir; requires either a species column, or that all rows are a single species)</li>
</ul>
</li>
<li>optionally: filter both files to the desired sample sets (e.g. filter on species, check common sample identifiers exist)</li>
<li>pass filtered genotype &amp; phenotype objects (which have common sample identifiers) to functions for
<ul>
<li>generating binary matrix of SIR vs marker presence/absence suitable for regression modelling (<code>get_binary_matrix</code>)</li>
<li>cross-tabulating SIR vs marker presence/absence, calculating &amp; plotting PPV (<code>solo_ppv_analysis</code>)</li>
<li>upset plots showing MIC/DD distribution stratified by genotype profile (<code>amr_complexUpset</code> or <code>amr_upset</code>)</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="code-for-testing-harmonize_data---but-note-the-required-input-files-are-not-in-this-repo">Code for testing harmonize_data - but note the required input files are not in this repo<a class="anchor" aria-label="anchor" href="#code-for-testing-harmonize_data---but-note-the-required-input-files-are-not-in-this-repo"></a>
</h3>
<pre><code><span><span class="co"># test code amrfinder plus</span></span>
<span><span class="co"># note both amrfinderplus test files appear malformed according to hamronize</span></span>
<span><span class="co"># and produce errors but the code works</span></span>
<span><span class="va">user_software_name</span> <span class="op">&lt;-</span> <span class="st">"amrfinderplus"</span></span>
<span><span class="va">user_software_version</span> <span class="op">&lt;-</span> <span class="st">"3.12.8"</span></span>
<span><span class="va">user_input_filename</span> <span class="op">&lt;-</span> <span class="st">"/Users/lshzd1/Desktop/ATB_Achromobacter_AFP.tsv"</span></span>
<span><span class="va">user_database_version</span> <span class="op">&lt;-</span> <span class="st">"2024-01-31.1"</span></span>
<span></span>
<span><span class="va">test_data</span> <span class="op">&lt;-</span> <span class="fu">harmonize_data</span><span class="op">(</span><span class="va">user_software_name</span>, <span class="va">user_software_version</span>,</span>
<span>                            <span class="va">user_database_version</span>, <span class="va">user_input_filename</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># test code rgi - same arguments as for amrfinder plus</span></span>
<span><span class="va">user_software_name</span> <span class="op">&lt;-</span> <span class="st">"rgi"</span></span>
<span><span class="va">user_software_version</span> <span class="op">&lt;-</span> <span class="st">"version x"</span></span>
<span><span class="va">user_input_filename</span> <span class="op">&lt;-</span> <span class="st">"/Users/lshzd1/Desktop/2025-01-14_11:06:52.908_KPN2009.fasta.txt"</span></span>
<span><span class="va">user_database_version</span> <span class="op">&lt;-</span> <span class="st">"database y"</span></span>
<span></span>
<span><span class="va">test_data</span> <span class="op">&lt;-</span> <span class="fu">harmonize_data</span><span class="op">(</span><span class="va">user_software_name</span>, <span class="va">user_software_version</span>,</span>
<span>                            <span class="va">user_database_version</span>, <span class="va">user_input_filename</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># test code - resfinder - must be json (hamronize can't do txt file)</span></span>
<span><span class="va">user_software_name</span> <span class="op">&lt;-</span> <span class="st">"resfinder"</span></span>
<span><span class="va">user_software_version</span> <span class="op">&lt;-</span> <span class="st">"4.6.0"</span></span>
<span><span class="va">user_input_filename</span> <span class="op">&lt;-</span> <span class="st">"/Users/lshzd1/Desktop/KPN2214.json"</span></span>
<span><span class="va">user_database_version</span> <span class="op">&lt;-</span> <span class="st">"2024-08-06"</span></span>
<span></span>
<span><span class="va">test_data</span> <span class="op">&lt;-</span> <span class="fu">harmonize_data</span><span class="op">(</span><span class="va">user_software_name</span>, <span class="va">user_software_version</span>,</span>
<span>                            <span class="va">user_database_version</span>, <span class="va">user_input_filename</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/interpretAMR/AMRgen/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a> | file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing AMRgen</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Matthijs S. Berends <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-7620-1800" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Natacha Couto <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-9152-5464" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Jane Hawkey <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-9661-5293" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Kat Holt <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-3949-2471" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Matthijs S. Berends, Natacha Couto, Jane Hawkey, Kat Holt.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
